lxd/task/task.go

```diff
@@ -0,0 +1,82 @@
+package task
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+// A task executes a certain function periodically, according to a certain
+// schedule.
+type task struct {
+	f        Func          // Function to execute.
+	schedule Schedule      // Decides if and when to execute f.
+	reset    chan struct{} // Resets the shedule and starts over.
+}
+
+// Execute the our task function according to our schedule, until the given
+// context gets cancelled.
+func (t *task) loop(ctx context.Context) {
+	// Kick off the task immediately (as long as the the schedule is
+	// greater than zero, see below).
+	delay := immediately
+
+	for {
+		var timer <-chan time.Time
+
+		schedule, err := t.schedule()
+		switch err {
+		case ErrSkip:
+			// Reset the delay to be exactly the schedule, so we
+			// rule out the case where it's set to immediately
+			// because it's the first iteration or we got reset.
+			delay = schedule
+			fallthrough // Fall to case nil, to apply normal non-error logic
+		case nil:
+			// If the schedule is greater than zero, setup a timer
+			// that will expire after 'delay' seconds (or after the
+			// schedule in case of ErrSkip, to avoid triggering
+			// immediately), otherwise setup a timer that will
+			// never expire (hence the task function won't ever be
+			// run, unless Reset() is called and schedule() starts
+			// returning values greater than zero).
+			if schedule > 0 {
+				timer = time.After(delay)
+			} else {
+				timer = make(chan time.Time)
+			}
+		default:
+			// If the schedule is not greater than zero, abort the
+			// taks and return immediately. Otherwise set up the
```

Comment:
task

Reply:
Fixed in f8e6df366c5a18ba7dcbd0a3a72e05023162fc3d too.